browser/src/control/jsdialog/Util.Accessibility.ts

```diff
@@ -0,0 +1,118 @@
+/* -*- js-indent-level: 8 -*- */
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * JSDialog.Accessibility - accessibility utilities for JSDialog widgets
+ */
+
+declare var JSDialog: any;
+
+function findLabelElementById(
+	container: HTMLElement | Document,
+	labelledById: string,
+	suffix: string,
+): HTMLElement | null {
+	return (
+		container.querySelector(`#${labelledById + '-label-' + suffix}`) ||
+		container.querySelector(`#${labelledById + '-label'}`) ||
+		container.querySelector(`#${labelledById}`)
+	);
+}
+
+function setupA11yLabelForLabelableElement(
+	parentContainer: HTMLElement,
+	content: HTMLElement,
+	data: WidgetJSON,
+	builder: JSBuilder,
+) {
+	app.layoutingService.appendLayoutingTask(function () {
+		app.layoutingService.appendLayoutingTask(function () {
+			if (!data.labelledBy) {
+				JSDialog.AddAriaLabel(content, data, builder);
+				return;
+			}
+
+			const element =
+				findLabelElementById(
+					parentContainer,
+					data.labelledBy,
+					builder.options.suffix,
+				) ||
+				findLabelElementById(document, data.labelledBy, builder.options.suffix);
+
+			if (!element) {
+				JSDialog.AddAriaLabel(content, data, builder);
+				return;
+			}
+
+			const labelHasHtmlFor =
+				element.tagName === 'LABEL' && (element as HTMLLabelElement).htmlFor;
+
+			const htmlForPointsToThisElement =
+				labelHasHtmlFor && (element as HTMLLabelElement).htmlFor === content.id;
+
+			if (!htmlForPointsToThisElement) {
+				content.setAttribute('aria-labelledby', element.id);
+			}
+		});
+	});
+}
+
+function setupA11yLabelForNonLabelableElement(
+	container: HTMLElement,
+	data: WidgetJSON,
+	builder: JSBuilder,
+) {
+	if (data.labelledBy)
+		container.setAttribute('aria-labelledby', data.labelledBy);
+	else JSDialog.AddAriaLabel(container, data, builder);
+}
+
+function addAriaLabel(
+	element: HTMLElement,
+	data: WidgetJSON,
```

Comment:
I would say -> here we should use the "read id" from the element which contains label
so `findLabelElementById` to get element -> then use it's id as a value, so we are sure it exists